Add a new section detailing the various potential realms

domenic opened the pull request:
This adds a new section under scripts detailing the entry, incumbent,
current, and relevant concepts with regard to realms/global objects/
environment settings objects. This centralizes information that was
previously somewhat spread out, and adds lots of details, as well as
an example.

This takes care of some of the pain points noted in #473, in particular
fixing the definition of entry settings object to be more clearly
defined in terms of the JavaScript execution context stack. There is
still some work to do in that issue: namely, updating Web IDL with
regard to tracking the entry settings object, and also making sure that
incumbent settings object is correctly defined, which I haven't yet tackled.

This closes #167 by finally implementing the plan in
https://github.com/whatwg/html/issues/167#issuecomment-195478361 for
making the correspondences clear.

/cc @bzbarsky; your review would be much appreciated.

https://dl.dropboxusercontent.com/u/20140634/realms-globals-settings/index.html contains a rendered version. In particular:

- [New section centralizing all the information](https://dl.dropboxusercontent.com/u/20140634/realms-globals-settings/index.html#realms,-settings-objects,-and-global-objects)
- [Promise job handling](https://dl.dropboxusercontent.com/u/20140634/realms-globals-settings/index.html#enqueuejob(queuename,-job,-arguments)) (slightly tweaked)

annevk commented on source:
```diff
@@ -87502,7 +87467,8 @@ interface <dfn>NavigatorOnLine</dfn> {
 
-   <li><p>Label <var>settings</var> as a <span>candidate entry settings object</span>.</p></li>
+   <li><p>Set <var>settings</var>'s <span>realm execution context</span>'s <span>is candidate entry
+   execution context</span> flag to true.</p></li>
```
Either set/unset an "x flag", or set "x" to true/false.

annevk commented on source:
```diff
@@ -87543,18 +87509,205 @@ interface <dfn>NavigatorOnLine</dfn> {
+
+  <p>When the user agent is to <dfn>run the global script clean-up jobs</dfn>, the user agent must
+  perform each of the jobs in the <span>global script clean-up jobs list</span> and then empty the
+  list.</p>
```
This doesn't really define what the list contains and why the items in the list can be performed (run might be clearer or a mention of using ECMAScript terminology, though then jobs might be confusing?).

domenic replied:
This just moves the existing text here; this patch is unrelated to the global script clean-up jobs.

annevk commented on source:
```diff
@@ -87543,18 +87509,205 @@ interface <dfn>NavigatorOnLine</dfn> {
+  <p>All <span data-x="JavaScript execution context">JavaScript execution contexts</span> must
+  contain a boolean flag, <dfn>is candidate entry execution context</dfn>, which is initially set
+  to false. In the process of <a href="#calling-scripts">calling scripts</a>, this flag can be
+  toggled.</p>
```
So here we are extending an ECMAScript concept without using a [[HostDefined]] slot, right? I think that might be worth explaining.

domenic replied:
Yeah, probably. Execution contexts aren't records or objects with slots, but instead have

> code evaluation state: Any state needed to perform, suspend, and resume evaluation of the code associated with this execution context.

I'll add a patch explaining that this is part of the code evaluation state

annevk commented:
Really glad you are sorting this out, even with examples \o/

@bholley might want to have a look too I think.

bholley commented on source:
```diff
@@ -87543,18 +87509,205 @@ interface <dfn>NavigatorOnLine</dfn> {
+  new specifications, and we are considering whether it can be removed from the platform. See <a
+  href="https://www.w3.org/Bugs/Public/show_bug.cgi?id=26603">Bugzilla bug 26603</a>. Currently, the
+  <span data-x="concept-incumbent-everything">incumbent</span> concept is used in some security
+  checks.</p>
```
Per https://www.w3.org/Bugs/Public/show_bug.cgi?id=26603#c36 I think we're stuck with the incumbent, so it isn't really accurate to suggest that it can be removed. But we should continue to discourage people from using it.

bholley commented:
I kinda skimmed, but it looks great in general - thanks for clarifying all this stuff!

domenic commented:
Great! One thing I wasn't entirely clear on is whether entry is a legacy concept or not. Some bugs seem to indicate that we shouldn't use it for origin checks, but that's not necessarily the same as saying we shouldn't use it at all. But in https://www.w3.org/Bugs/Public/show_bug.cgi?id=26603#c41, you say

> The entry global is basically only relevant for legacy reasons. The incumbent might be more objectively necessary, but only for very specific use-cases (the only one I can think of is referrer). So I think we should basically discourage the use of both.

So I might want to tweak the wording. Our recommendation is to use "current" in most cases, except to use "relevant" in the particular case explained here (i.e. creating an object which is cached)?

domenic commented:
Addressed review feedback, and also expanded on the relevant example (with getBattery()). I've assumed that we want to recommend against both entry and incumbent going forward, and tell people to use either current or relevant.

I've updated the rendered version so the links in the OP will reflect these latest changes.

bholley commented:
Yeah, we should discourage entry unless there's a very good specific reason. Current and Relevant should suffice in almost all cases.

Ms2ger commented on source:
```diff
@@ -87527,18 +87493,226 @@ interface <dfn>NavigatorOnLine</dfn> {
+   <p>Each page has its own <span>browsing context</span>, and thus its own <span>JavaScript
+   realm</span>, <span>global object</span>, and <span>settings object</span>.</p>
+
+   <p>When the button is pressed in <code data-x="">a.html</code>, then:</p>
```
I think "When the `print` function is called in response to pressing the button in …" would be more accurate.

Ms2ger commented on source:
```diff
@@ -87527,18 +87493,226 @@ interface <dfn>NavigatorOnLine</dfn> {
+&lt;title>Relevant Realm demo: outer page&lt;/title>
+&lt;script>
+  function doTest() {
+    const promise = navigator.getBattery().call(frames[0].navigator);
```
No parens after `getBattery`

Ms2ger commented:
Thanks, this is great!

domenic commented:
Fixed @Ms2ger's helpful finds.

bzbarsky commented on source:
```diff
@@ -3320,6 +3320,17 @@ a.setAttribute('href', 'http://example.com/'); // change the content attribute d
 
    </dd>
 
+   <dt>Battery Status</dt>
```
This seems unrelated to the various realms thing...

bzbarsky commented on source:
```diff
@@ -87506,18 +87472,227 @@ interface <dfn>NavigatorOnLine</dfn> {
+  <dl>
+    <dt><dfn data-x="concept-entry-everything">Entry</dfn></dt>
+    <dd>This corresponds to the script that initiated the currently running script action: i.e.,
+    the script that was on the stack when the user agent called in to author code.</dd>
```
Probably better to phrase it as "the function or script that the user agent called into when it called into author code"?

domenic replied:
This and the other descriptions are fairly informal, so your suggestions for making them more precise are appreciated. I'll incorporate them now.

bzbarsky commented on source:
```diff
@@ -87506,18 +87472,227 @@ interface <dfn>NavigatorOnLine</dfn> {
+    the script that was on the stack when the user agent called in to author code.</dd>
+
+    <dt><dfn data-x="concept-incumbent-everything">Incumbent</dfn></dt>
+    <dd>This corresponds to the most-recently-entered author script on the stack.</dd>
```
Just to check, does "script" include "function"?  That's not entirely obvious to me... and it should, for this definition to make sense.

bzbarsky commented on source:
```diff
@@ -87506,18 +87472,227 @@ interface <dfn>NavigatorOnLine</dfn> {
+
+    <dt><dfn data-x="concept-current-everything">Current</dfn></dt>
+    <dd>This corresponds to the currently-running function object, including built-in user-agent
+    functions which might not be implemented as JavaScript.</dd>
```
Might make sense to point to the idea of current Realm here.

bzbarsky commented on source:
```diff
@@ -87506,18 +87472,227 @@ interface <dfn>NavigatorOnLine</dfn> {
+  concept is used in some security checks, and the <span
+  data-x="concept-entry-everything">entry</span> concept is used to obtain, amongst other things,
+  the <span>API base URL</span> to <span data-x="parse a url">parse a URL</span>, used in scripts
+  running in that <span>unit of related similar-origin browsing contexts</span>.</p>
```
But only in some APIs is entry used.  In others, the relevant Realm is used....  Not sure that's worth mentioning here.

domenic replied:
This is lifted from existing text. I'll make it "sometimes used" instead of just "used".

bzbarsky commented on source:
```diff
@@ -87506,18 +87472,227 @@ interface <dfn>NavigatorOnLine</dfn> {
+  context</dfn> flag, which is initially unset. In the process of <a href="#calling-scripts">calling
+  scripts</a>, this flag will be toggled.</p>
+
+  <p>With this in hand, we define the <dfn>entry execution context</dfn> to be the topmost entry in
```
This assumes a stack that grows "up".  Is that a reasonable assumption, or one spelled out somewhere?

domenic replied:
Yes; it's in https://tc39.github.io/ecma262/#sec-execution-contexts, albeit slightly implicit. I'll change it to "most recently pushed"

bzbarsky commented:
Thank you for writing this down!

domenic commented:
Pushed a commit and updated the live version with @bzbarsky's fixes. Still some discussion going on in https://github.com/whatwg/html/issues/473#issuecomment-214539307, but it might be time to merge this and work on further follow-ups in another PR? What do people think?

domenic commented:
Per offthread discussions with @bzbarsky I need to make it clearer that "relevant" is a predicate that applies to objects, in contrast to current/incumbent/entry which are ambient values. That is, you can have "relevant settings object of X" and "relevant settings object of Y". In most algorithms it's "the relevant settings object of **this**" that matters.

domenic commented:
Updated; the last two commits should address @bzbarsky's concerns.

annevk commented:
@bzbarsky any remaining concerns or can I go ahead and merge this?

bzbarsky commented:
No remaining concerns, I think.
